src/org/sosy_lab/java_smt/solvers/princess/PrincessRationalFormulaManager.java

```diff
@@ -0,0 +1,136 @@
+      NonLinearArithmetic pNonLinearArithmetic,
+      PrincessIntegerFormulaManager pIntegerFormulaManager) {
+    super(pCreator, pNonLinearArithmetic);
+    this.ifmgr = pIntegerFormulaManager;
+  }
+
+  @Override
+  protected boolean isNumeral(IExpression value) {
+    if (value instanceof IFunApp) {
+      IFunApp fun = (IFunApp) value;
+      switch (fun.fun().name()) {
+        case "int":
+        case "Rat_int":
+          assert fun.fun().arity() == 1;
+          return ifmgr.isNumeral(fun.apply(0));
+        case "frac":
+          assert fun.fun().arity() == 2;
+          return ifmgr.isNumeral(fun.apply(0)) && ifmgr.isNumeral(fun.apply(1));
+      }
+    }
+    return false;
+  }
+
+  protected IExpression fromInteger(ITerm i) {
+    return PrincessEnvironment.rationalTheory.int2ring(i);
+  }
+
+  @Override
+  protected IExpression makeNumberImpl(long i) {
+    return fromInteger(ifmgr.makeNumberImpl(i));
+  }
+
+  @Override
+  protected IExpression makeNumberImpl(BigInteger i) {
+    return fromInteger(ifmgr.makeNumberImpl(i));
+  }
+
+  @Override
+  protected IExpression makeNumberImpl(Rational pI) {
+    return divide(makeNumberImpl(pI.getNum()), makeNumberImpl(pI.getDen()));
+  }
+
+  @Override
+  protected IExpression makeNumberImpl(String i) {
+    return makeNumberImpl(new BigDecimal(i));
+  }
+
+  @Override
+  protected IExpression makeNumberImpl(double pNumber) {
+    return makeNumberImpl(BigDecimal.valueOf(pNumber));
+  }
+
+  @Override
+  protected IExpression makeNumberImpl(BigDecimal pNumber) {
+    if (pNumber.stripTrailingZeros().scale() <= 0) {
+      // We have an integer number
+      // Return the term for a/1
+      return PrincessEnvironment.rationalTheory.int2ring(
+          ifmgr.makeNumberImpl(pNumber.toBigInteger()));
+    } else {
+      // We have a fraction a/b
+      // Convert the numerator and the divisor and then return the fraction
+      List<ITerm> args =
+          ImmutableList.of(
+              ifmgr.makeNumberImpl(pNumber.unscaledValue()),
+              ifmgr.makeNumberImpl(BigInteger.valueOf(10).pow(pNumber.scale())));
+      return new IFunApp(
+          PrincessEnvironment.rationalTheory.frac(), PrincessEnvironment.toSeq(args));
+    }
+  }
+
+  @Override
+  protected IExpression makeVariableImpl(String varName) {
+    return getFormulaCreator().makeVariable(PrincessEnvironment.FRACTION_SORT, varName);
+  }
+
+  @Override
+  protected IExpression floor(IExpression number) {
+    throw new AssertionError("floor is not supported in Princess");
+  }
+
+  @Override
+  protected IExpression multiply(IExpression number1, IExpression number2) {
+    FormulaType<?> sort1 = getFormulaCreator().getFormulaType(number1);
+    FormulaType<?> sort2 = getFormulaCreator().getFormulaType(number1);
+
+    IExpression result = PrincessEnvironment.rationalTheory.mul((ITerm) number1, (ITerm) number2);
+
+    if (result instanceof IIntLit && ((IIntLit) result).value().equals(IdealInt.apply(0))) {
```

Comment:
A comment for explaining this line would be nice.

Reply:
see next line 😄